File: pset2/readabillity/readability.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <cs50.h>

// prototypes
int count_letters(string prompt);
int count_words(string prompt);
int count_sentences(string prompt);

int main(void)
{
    string prompt = get_string("Text: ");

    float total_letters = count_letters(prompt);
    float total_words = count_words(prompt);
    float total_sentences = count_sentences(prompt);

    // L is the average number of letters per 100 words
    float L = (total_letters / total_words) * 100;
    // S is the average number of sentences per 100 words
    float S = (total_sentences / total_words) * 100;

    float index = 0.0588 * L - 0.296 * S - 15.8;

    /*
    debug
    printf("Letters:  %f\n", total_letters);
    printf("Words:  %f\n", total_words);
    printf("Sentences:  %f\n", total_sentences);
    printf("L:  %f\n", L);
    printf("S:  %f\n", S);
    */

    // print the grade
    if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %.f\n", index);
    }

}

// count letters from the input prompt
int count_letters(string prompt)
{
    int letter_count = 0;
    // get the total length of the string
    for (int i = 0, text_count = strlen(prompt); i < text_count; i++)
    {
        // filter out non alphabetical characters
        if (isalpha(prompt[i]))
        {
            letter_count ++;
        }
    }
    return letter_count;
}

// count breaks in alphabetic chars to infer words
int count_words(string prompt)
{
    int word_count = 1;
    // get the total length of the string
    for (int i = 0, text_count = strlen(prompt); i < text_count; i++)
    {
        // count text as a word when unbroken by non alphabetical chars.
        if (isspace(prompt[i]))
        {
            word_count ++;
        }
    }
    return word_count;
}

// count '.', '!' and '?' as sentence boundaries
int count_sentences(string prompt)
{
    int sentence_count = 0;
    // get the total length of the string
    for (int i = 0, text_count = strlen(prompt); i < text_count; i++)
    {
        // count a sentence when encountering the following ASCII chars.
        if ((prompt[i] == '.') || (prompt[i] == '!') || (prompt[i] == '?'))
        {
            sentence_count ++;
        }
    }
    return sentence_count;
}

File: pset6/readability/readability.py
import re

from cs50 import get_string

# readability.py


def main():
    """ execute program """
    string = get_string("Text: ")
    calc(string)


def calc(string):
    """ calculate readability usng Coleman-Liau formula """

    total_letters = countLetters(string)
    total_words = countWords(string)
    total_sentences = countSentences(string)

    # average number of letters per 100 words
    L = (total_letters / total_words) * 100
    # average number of sentences per 100 words
    S = (total_sentences / total_words) * 100

    index = 0.0588 * L - 0.296 * S - 15.8

    # print out grade
    if index > 16:
        print("Grade 16+")
    elif index < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {index:.0f}")


def countLetters(string):
    """ count a-z and A-Z as letters """
    letter_count = 0
    for letter in string:
        if letter.isalpha():
            letter_count += 1

    return letter_count


def countWords(string):
    """ count words using .split() """
    word_count = len(string.split())
    return word_count


def countSentences(string):
    """ count sentences by finding '.', '!' and '?' as sentence boundaries """
    sentence_count = 0
    for letter in string:
        if letter in [".", "!", "?"]:
            sentence_count += 1

    return sentence_count


if __name__ == "__main__":
    main()